src/foam/dao/PipelinePMDAO.java

```diff
@@ -0,0 +1,135 @@
+    findName_      = getOf().getId() + ":" + delegateName_ + ":pipeFind";
+    removeName_    = getOf().getId() + ":" + delegateName_ + ":pipeRemove";
+    removeAllName_ = getOf().getId() + ":" + delegateName_ + ":pipeRemoveAll";
+  }
+
+  /** Creates the PM pipeline by adding an EndPipelinePMDAO after of this class only if it is a ProxyDAO. 
+  *   If the delegate of that is also a ProxyDAO, creates a new PipelinedPMDAO in the chain beofre it which repeats this procedure recursively. */
+  void createPipeline() {
+    removePipelineFragmentation();
+    DAO delegate = getDelegate();
+    DAO secondaryDelegate;
+    secondaryDelegate = ((ProxyDAO) delegate).getDelegate();
+    ((ProxyDAO) delegate).setDelegate(new EndPipelinePMDAO(getX(), secondaryDelegate));
+    delegate = ((ProxyDAO) delegate).getDelegate();
+    if ( secondaryDelegate instanceof ProxyDAO ) {
+      ((ProxyDAO) delegate).setDelegate(new PipelinePMDAO(getX(), secondaryDelegate));
+    }
+  }
+
+  /** Creates the PM that will measure the performance of each operation and creates a new context with it as a variable which the EndPipelinePMDAO
+   *  will use to access the pm after it is passed onto it through the arguments of the DAO operations */
+  X createPMX(String name, X x) {
+    PM pm = new PM();
+    pm.setClassType(PipelinePMDAO.getOwnClassInfo());
+    pm.setName(name);
+    return x.put("pipePmStart", pm);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    return super.put_(createPMX(putName_, x), obj);
+  }
+
+  @Override
+  public FObject find_(X x, Object id) {
+    return super.find_(createPMX(findName_, x), id);
+  }
+
+  @Override
+  public FObject remove_(X x, FObject obj) {
+    return super.remove_(createPMX(removeName_, x), obj);
+  }
+
+  @Override
+  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
+    super.removeAll_(createPMX(removeAllName_, x), skip, limit, order, predicate);
+  }
+
+  class EndPipelinePMDAO extends ProxyDAO {
+    public EndPipelinePMDAO(X x, DAO delegate) {
+      super(x, delegate);
+    }
+
+    @Override
+    public FObject put_(X x, FObject obj) {
+      PM pm = (PM) x.get("pipePmStart");
+      if ( pm != null )
+        pm.log(x);
+      return super.put_(x, obj);
+    }
+
+    @Override
+    public FObject find_(X x, Object id) {
+      PM pm = (PM) x.get("pipePmStart");
+      if ( pm != null )
+        pm.log(x);
+      return super.find_(x, id);
+    } 
+
+    @Override
+    public FObject remove_(X x, FObject obj) {
+      PM pm = (PM) x.get("pipePmStart");
+      if ( pm != null )
+        pm.log(x);
+      return super.remove_(x, obj);
+    }
+
+    @Override
+    public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
+      PM pm = (PM) x.get("pipePmStart");
+      if ( pm != null )
+        pm.log(x);
+      super.removeAll_(x, skip, limit, order, predicate);
+    }
+  }
+
+  void removePipelineFragmentation() {
```

Comment:
What does this do?

Reply:
It essentially just cleans up any "junk" pipePMDAO decorators to prevent double layering of pipePMDAO's. Say you decorate account, then localAccount is also decorated as a result and then if you decorate localAccount you are essentially doubling down except on the initial decorator.

Reply:
I don't think this is needed now that you added the check to see if the delegate is a PMPipelineDAO.